src/measure_theory/integral/layercake.lean

```diff
@@ -0,0 +1,231 @@
+  { refine λ t t_pos, ⟨integrable_on.congr_fun' (g_intble t t_pos).1 (g_eq_G_on t), _⟩,
+    rw Ioc_eq_empty_of_le t_pos.lt.le,
+    exact integrable_on_empty, },
+  have eq₁ : ∫⁻ t in Ioi 0, μ {a : α | t ≤ f a} * ennreal.of_real (g t)
+             = ∫⁻ t in Ioi 0, μ {a : α | t ≤ f a} * ennreal.of_real (G t),
+  { apply lintegral_congr_ae,
+    filter_upwards [g_eq_G] with a ha,
+    rw [ha], },
+  have eq₂ : ∀ ω, ∫ t in 0..f ω, g t = ∫ t in 0..f ω, G t,
+  { refine λ ω, interval_integral.integral_congr_ae _,
+    have fω_nn : 0 ≤ f ω := f_nn ω,
+    rw [interval_oc_of_le fω_nn,
+      ← ae_restrict_iff' (measurable_set_Ioc : measurable_set (Ioc (0 : ℝ) (f ω)))],
+    exact g_eq_G_on (f ω), },
+  simp_rw [eq₁, eq₂],
+  exact layercake_of_measurable μ f_nn f_mble G_intble G_mble (λ t t_pos, G_nn t),
+end
+
+theorem layercake_one (μ : measure α) [sigma_finite μ] (f_nn : 0 ≤ f) (f_mble : measurable f) :
+  ∫⁻ ω, ennreal.of_real (f ω) ∂μ = ∫⁻ t in Ioi 0, (μ {a : α | t ≤ f a}) :=
+begin
+  set cst := λ (t : ℝ), (1 : ℝ) with def_cst,
+  have cst_intble : ∀ t > 0, interval_integrable cst volume 0 t,
+    from λ _ _, interval_integrable_const,
+  have key := layercake μ f_nn f_mble cst_intble (eventually_of_forall (λ t, zero_le_one)),
+  simp_rw [def_cst, ennreal.of_real_one, mul_one] at key,
+  rw ← key,
+  congr' with ω,
+  simp only [interval_integral.integral_const, sub_zero, algebra.id.smul_eq_mul, mul_one],
+end
+
+-- TODO: The assumption `(p_large : 0 < p)` should suffice, but the assumptions of
+-- `interval_integral.interval_integrable_rpow` and `interval_integral.integral_rpow`
+-- are too stringent to apply here.
+theorem layercake_p (μ : measure α) [sigma_finite μ]
+  (f_nn : 0 ≤ f) (f_mble : measurable f) {p : ℝ} (p_large : 1 < p) :
+  ∫⁻ ω, ennreal.of_real ((f ω)^p) ∂μ
+    = (ennreal.of_real p) * ∫⁻ t in Ioi 0, (μ {a : α | t ≤ f a}) * ennreal.of_real (t^(p-1)) :=
+begin
+  have p_pos : 0 < p := zero_le_one.trans_lt p_large,
+  have p_sub_one : 0 ≤ p - 1 := by linarith,
+  have obs : ∀ (x : ℝ), (∫ (t : ℝ) in 0..x, t^(p-1)) = x^p / p,
+  { intros x,
+    rw integral_rpow (or.inl p_sub_one),
+    simp [real.zero_rpow p_pos.ne.symm], },
+  set g := λ (t : ℝ), t^(p-1) with g_def,
+  have g_nn : ∀ᵐ t ∂(volume.restrict (Ioi (0 : ℝ))), 0 ≤ g t,
+  { filter_upwards [self_mem_ae_restrict (measurable_set_Ioi : measurable_set (Ioi (0 : ℝ)))],
+    intros t t_pos,
+    rw g_def,
+    exact real.rpow_nonneg_of_nonneg (mem_Ioi.mp t_pos).le (p - 1), },
+  have g_intble : ∀ t > 0, interval_integrable g volume 0 t,
+    from λ _ _, interval_integral.interval_integrable_rpow (or.inl p_sub_one),
+  have key := layercake μ f_nn f_mble g_intble g_nn,
+  simp_rw [g_def] at key,
+  rw [← key, ← lintegral_const_mul (ennreal.of_real p)]; simp_rw obs,
+  { congr' with ω,
+    rw [← ennreal.of_real_mul p_pos.le, mul_div_cancel' ((f ω)^p) p_pos.ne.symm], },
+  { exact ((f_mble.pow measurable_const).div_const p).ennreal_of_real, },
```

Comment:
This one-line proof that replaces the bloc you had before was found by `measurability?`.